lib/using_restricted_array.rb

```diff
@@ -6,56 +6,124 @@
+# Space complexity: constant, there are 2 variables created but they are over-written and do not depend on the size of the array
 def length(array)
-  raise NotImplementedError
+  i = 0
+  length = 0
+  21.times do
+    until array[i] == nil
+      i += 1
+      length += 1
+    end
+  end
+  return length
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear, please see explanation on LN 9. 
+# Space complexity: constant, no variables are stored in memory.
 def print_array(array)
-  raise NotImplementedError
+  21.times do
+    until array[i] == nil
+      i += 1
+      puts array[i]
+    end
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear, the array is passed through once, as the array grows, the time required to look through each value increases at the same rate.
+# Space complexity: contant, the one stored value is over-written and does not depend on the size of the array
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  result = false
+  i = 0
+  length.times do
+    i += 1
+    if array[i] == value_to_find
+      result = true
+    end
+  end
+  return result
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear, the array is passed through once, as the array grows, the time required to look through each value increases at the same rate.
+# Space complexity: constant, there are 2 variables created but they are over-written and do not depend on the size of the array
 def find_largest(array, length)
-  raise NotImplementedError
+  high_value = array[0]
+  i = 0
+  (length).times do 
+    if array[i] >= high_value
+      high_value = array[i]
+    end
+    i += 1
+  end
+  return high_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear, the array is passed through once, as the array grows, the time required to look through each value increases at the same rate.
+# Space complexity: constant, there are 2 variables created but they are over-written and do not depend on the size of the array
 def find_smallest(array, length)
-  raise NotImplementedError
+  low_value = array[0]
+  i = 0
+  (length).times do 
+    if array[i] <= low_value
+      low_value = array[i]
+    end
+    i += 1
+  end
+  return low_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: quadratic, the array is iterated through twice, so as it gorws, the time required to process it would grow twice as much.
```

Comment:
Since the loops are not nested, and each loop runs n times.  The time complexity is O(n)